core/src/banking_stage.rs

```diff
@@ -1864,57 +1920,52 @@ impl BankingStage {
         process_transactions_summary
     }
 
+    // Returns the number of packets that were filtered out for
+    // no longer being valid (could be too old, a duplicate of something
+    // already processed, etc.)
     fn filter_unprocessed_packets_at_end_of_slot(
-        bank: &Arc<Bank>,
-        deserialized_packet_batch: &HashMap<usize, DeserializedPacket>,
+        bank: &Option<Arc<Bank>>,
+        unprocessed_packets: &mut UnprocessedPacketBatches,
         my_pubkey: &Pubkey,
         next_leader: Option<Pubkey>,
         banking_stage_stats: &BankingStageStats,
-    ) -> Vec<usize> {
+    ) -> usize {
         // Check if we are the next leader. If so, let's not filter the packets
         // as we'll filter it again while processing the packets.
         // Filtering helps if we were going to forward the packets to some other node
-        if let Some(leader) = next_leader {
-            if leader == *my_pubkey {
-                return deserialized_packet_batch
-                    .keys()
-                    .cloned()
-                    .collect::<Vec<usize>>();
-            }
+        let will_still_be_leader = next_leader
+            .map(|next_leader| next_leader == *my_pubkey)
+            .unwrap_or(false);
+        let should_filter_unprocessed_packets = !will_still_be_leader && bank.is_some();
+        let original_unprocessed_packets_len = unprocessed_packets.len();
+
+        if should_filter_unprocessed_packets {
+            // If `should_filter_unprocessed_packets` is true, then the bank
+            // must be `Some`
+            let bank = bank.as_ref().unwrap();
+            let mut unprocessed_packet_conversion_time =
+                Measure::start("unprocessed_packet_conversion");
+
+            let should_retain = |deserialized_packet: &DeserializedPacket| {
+                Self::transaction_from_deserialized_packet(
+                    deserialized_packet.immutable_section(),
+                    &bank.feature_set,
+                    bank.vote_only_bank(),
+                    bank.as_ref(),
+                )
+                .is_some()
+            };
+            unprocessed_packets.retain(|deserialized_packet| should_retain(deserialized_packet));
```

Comment:
```suggestion
            unprocessed_packets.retain(should_retain);
```

I think that works the same on closures as function calls

Reply:
fixed